[Step3] 다대다 연관관계 리팩토링 리뷰 요청드립니다.

ccf05017 opened the pull request:
안녕하세요, 우아한테크코스 Pro 수강생 조범진 입니다 :)

Step3 다대다 연관관계 리팩토링을 진행했습니다.
DM으로 말씀드린대로 이전 단계에서 미리 중간 매핑 테이블을 통한 다대다 관계를 진행해서
이번 단계에서는 ManyToMany를 통해 Line, Station 간 연관관계를 맺고 CRUD 테스트를 진행했습니다.

진행하면서 자동으로 생성되는 연관 관계 매핑 테이블에 간섭할 수  없어서 여러모로 제가 원하는
비즈니스 로직을 넣기 어렵다는 것을 배울 수 있었습니다.

관련해서 부족한 부분이나 개선할 필요가 있는 부분 있으면 피드백 부탁드립니다 :)
감사합니다!

ccf05017 committed:
docs(README): add requirements of step3

- 다대다 연관관계 관련 요구사항 정리

ccf05017 committed:
faet(ManyToMany): add entities and repositories

- ManyToMany 어노테이션을 활용한 다대다 매핑을 테스트하기 위해 기본 엔티티 및 리포지토리 생성

ccf05017 committed:
test(ManyToManyTest): save station when save related line

- ManyToMany 연관관계를 통해 Line 저장시 연관된 Station도 저장되는 것을 확인하는 테스트 작성

ccf05017 committed:
test(ManyToManyTest): get station at line

- ManyToMany 연관관계의 Line에서 연관된 Station 조회 가능 여부 확인

ccf05017 committed:
test(ManyToManyTest): get Line at Station

- ManyToMany 연관관계의 Station에서 Line으로 오브젝트 탐색이 가능한 것을 확인하는 테스트 작성

ccf05017 committed:
test(ManyToManyTest): dirty check test

- 영속화 된 엔티티의 더티체킹을 통한 자동 업데이트 기능 확인

ccf05017 committed:
test(ManyToManyTest): not delete station when delete related line

- 영속성 전이 속성 제한을 통해 Line 엔티티를 삭제해도 연관된 Staiton 엔티티가 남는지 확인

ccf05017 committed:
refactor(ManyToManyTest): extract method

- 테스트 코드 가독성을 높이기 위해 중복되서 사용되는 메서드 추출 및 의미 부여

ccf05017 committed:
refactor(ManyToManyTest): extract method

- 테스트 코드 가독성을 높이고 반복되는 코드를 줄이기 위한 메서드 추출

ccf05017 committed:
docs(README): update progress

- 진행상황 업데이트

young891221 commented on README.md:
```diff
@@ -121,3 +121,9 @@ create table favorite (
 - [X] 매핑시 불필요한 기본 옵션 제거
 - [X] 지나치게 장황한 테스트 메서드에서 공통 부분 추출하기
+
+## Step3. 다대다 연관 관계 리팩토링
```
이전 코딩베이스 기반이 아닌 완전히 새로 작성하셨네요. 이렇게 작성하신다면 distance에 대한 개념도 들어가야 될 것 같습니다ㅠㅠ
제가 이해한 방식은 이전 코드 기반에서 작성하는걸로 생각해서 해당 베이스를 바탕으로 의견을 드릴려고 했던 의도였습니다.
현재 버전에서 역과 역 사이의 거리 개념을 다룰 수 있도록 distance에 대한 개념을 추가해서 다시 PR 부탁드립니다.

ccf05017 replied:
아 제가 리뷰어님이 말씀하신 바를 완전히 잘못 이해했던 것 같습니다 ㅜ

저는 중간 매핑 테이블을 이용한 다대다 매핑은 이미 구현이 되어 있으니, 해당 사항을 ManyToMany 어노테이션으로 묶어서 처리해보라는 말씀으로 이해했습니다. 여기에 추가로 ManyToMany 관계에서는 중간 매핑 테이블에 속성을 더할 수 없기 때문에 distance를 추가할 수 없어서 이걸 실습으로 알아보라는 의미라고 혼자 해석했었습니다.

제가 알기론 ManyToMany의 경우 중간 매핑 테이블에 distance와 같이 제가 원하는 칼럼을 추가하는 방법은 없는 걸로 알고 있습니다.
혹시 의도하신 바가

1. ManyToMany 연관 관계를 사용하고 Line, Station 중 어느 한 곳에 distance 속성을 추가해서 매핑을 시도하라
2. ManyToMany 매핑에서 중간 관계 테이블을 쓰지 않고 distance 속성을 추가할 수 있는 방법이 있으니 이걸 찾아서 시도하라

둘 중 어떤 방향인지 알려주실 수 있을까요?

혹은 제가 여전히 방향을 잘못잡고 있다면 어떤 의도이신지 조금만 더 설명해주시면 감사하겠습니다 ㅜ 제약 조건이 반드시 ManyToMany 어노테이션을 활용해서 distance 속성을 추가하는 것인지, 아니면 다대다 매핑을 어떤 방법으로든 distance 속성을 추가해서 구현하면 되는 것인지 헷갈리네요 ㅜ

young891221 reviewed (changes requested):
안녕하세요, 범진님!
3단계 ManyToMany 방식을 잘 구현해 주셨네요 👍 
현재 완전히 새로운 클래스라 다시 코멘트 드렸습니다. 확인하시고 반영 부탁드릴게요:)

ccf05017 committed:
refactor(ManyToMany): remove useless

- 요구사항을 잘못 이해하고 구현한 잘못된 구성요소들 모두 제거

ccf05017 committed:
feat(Distance): add new entity

- Line에 종속적인 거리 Entity 구현
- 역간 거리와 역의 관계를 정의한다.
- 기존의 LineStation을 좀 더 역할에 걸맞는 Section으로 이름 변경
- Line에 종속적이도록 Section을 변경함에 따라 필요 없어진 LineStationRepository 제거

ccf05017 committed:
test(ManyToManyRefactorTest): can save distance by line

- Line을 추가하면서 역간 거리를 추가할 수 있는지 확인하는 테스트 진행

ccf05017 committed:
test(ManyToManyRefactorTest): can get stations at line

- 라인에 속한 역 정보들을 정상적으로 불러올 수 있는지 확인하는 테스트 진행

ccf05017 committed:
test(ManyToManyRefactorTest): add delete case

- 라인을 삭제하더라도 연관된 역이 삭제되지 않는지 확인하는 테스트 추가

ccf05017 committed:
refactor(ManyToManyRefactorTest): extract method

- 테스트 코드 가독성을 높이기 위한 메서드 추출

ccf05017 commented:
안녕하세요, 우아한테크캠프 Pro 수강생 조범진입니다 :)

제가 요구사항을 잘 이해 못하고 계속 질문을 드렸는데 상세하게 답변해주셔서 감사합니다.
피드백 주신 내용을 기반으로 제가 이해한 방향으로 다대다 관계를 다시 맺어서 리팩토링 진행했습니다.

제가 요구사항을 잘못 이해했던 건 크게 두가지 때문인 것 같습니다.
1. 명백히 제 잘못으로 어설프게 다대다 관계를 알고 안다고 착각해서 요구사항을 명확하게 분석하지 않았기 때문
2. 다음 과정(인수테스트)와 병행해서 풀면서 미묘하게 다른 두 도메인 간 요구사항을 잡아내지 못했기 때문

1번은 제가 앞으로 똑바로 처리해나가면 될 것 같지만, 2번에서 배운 게 많은 것 같습니다.
같은 도메인을 다루지만, 거리에 대한 개념을 구간으로 관심을 두냐 아니면 단순히 이전 역과의 거리로 두냐에 따라
도메인 해석 방법이 상상 이상으로 크게 달라지는 것 같습니다 ㅎㅎ..

선행 학습이 좋은 건 아니지만 다음 미션을 해두지 않으면 과정을 끝까지 소화하지 못할 것 같다는 압박감 때문에 다음 과제랑 자꾸 병행하게 되네요 ㅜ
차후부터는 좀 더 현재 진행중인 과제에 집중할 수 있도록 노력하겠습니다.

이번 리뷰도 잘 부탁드립니다 :)

young891221 commented on src/main/java/jpa/domain/Distance.java:
```diff
@@ -0,0 +1,47 @@
+    private int distance;
+
+    @OneToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
+    private Section section;
```
section, distance를 통해 관계를 푸셨네요 👍

young891221 replied:
Line, Station이 다대다 관계인데 `@ManyToMany`로 추가하여 관계를 저어보고 테스트도 작성해 보면 좋을것 같습니다.

ccf05017 replied:
으으으으음......... 이제 조금 리뷰어님 말씀하신 의도가 쪼끔씩 이해가 되는 것 같습니다!
제가 이해하는 게 좀 많이 느렸네요 ㅜㅜㅜㅜ

Distance는 말 그대로 역간의 거리(역 두개간의 관계에 대한 내용) 를 담당하는 오브젝트(Entity)로 생성하고,
Line, Station의 관계는 그와는 별개로 Line과 Station간의 관계로 ManyToMany 관계를 맺어보라는 말씀으로 이해가 됩니다.

제가 이해한 방향이 맞는지 확인 한번만 부탁드립니다 :)

young891221 reviewed (changes requested):
범진님, 미션 반영 감사합니다!
말씀드린데로 잘 반영해 주셨네요 💯 
@ManyToMany만 추가로 반영해 보면 어떨까 싶어 의견드렸습니다. 확인 부탁드려요~!

ccf05017 committed:
refactor(Line): add stations when add distance

- 데이너를 일관되게 관리하기 위해 역간 거리를 추가하면서 연관된 역 목록도 같이 업데이트 한다.

ccf05017 committed:
test(ManyToManyRefactorTest): get lines from stations

- 양방향 ManyToMany 관계를 통해 역에서 연관된 Line 정보를 조회할 수 잇다.

ccf05017 commented:
우선 제가 이해한대로 Line, Station 간 연관관계를 새로 맺어주고,
해당 관계를 통해 Line, Station간 서로 오브젝트를 탐색할 수 있도록 리팩토링 진행해 봤습니다.

변경 후 추가사항은 다음과 같습니다
- Line에서 distance를 추가할 때 관련된 역도 추가하도록 연관관계 편의 메서드 추가
- Line에서 관련된 역 목록 조회 시 distance 의존성 제거하고 순수한 Station과의 관계에 대해서 응답
- 테스트를 통해 새로 생성된 Line에서 Distance를 추가할 때 Station도 같이 추가되는지 확인
- 테스트를 통해 새로 Line을 생성하고 Distance를 추가했을 때 Station에서 관련된 Line 목록을 조회할 수 있는지 확인

그 외 업데이트, 삭제 테스트에서는 이미 Line과 Staiton 간의 관계에 대해서 검증을 하고 있어서 한꺼번에 확인이 됐습니다.

제가 잘 알지 못했던 부분이라 이번 요구사항까지 완료한 뒤에야 리뷰어님이 어떤 의도로 요구사항을 주셨는지 알 것 같았습니다.

Distance를 표현하기 위해서는 단순히 Line, Station의 관계가 아니라 Line, Station1, Station2의 관계가 필요했고 이 관계를 정의하기에는 이미 말씀주신대로 N x N 경우이기 때문에 단순 매핑으로 풀어내기 어려웠던 것 같습니다. 저는 단순히 Distance를 속성으로 봤는데, 아예 이 자체가 따로 관리되야 할 개념(도메인)이라 생각하고 처리하면서 관계를 훨씬 유연하게 풀 수 있었던 것 같습니다.

두번째로 제가 잘못 이해했던 점은 Distance에 이미 Station 관계가 표현됐다고 생각했는데, 정확하게는 Distance와 관련된 Station의 속성이지 Distance에서는 Line과 Station의 관계에 대한 정확한 정보를 찾을 수가 없는 상태였던 것 같습니다. 제가 구현했던 Line에서 Station에 있는 정보를 불러오는 기능은 단순히 Distance를 통해 계산된 값일 뿐이지 Line과 Station의 관계를 정확하게 나타낼 수 없었던 것 같습니다. 그 가장 큰 증거가 'Station에서 Line을 탐색할 방법이 없다' 였던 것 같습니다. 매핑이 제대로 되지 않았기 때문에 Station 입장에서 자기가 어디에 소속되어 있는지 어떻게 사용됐는지 전달받을 방법이 없었습니다.

피드백 주신대로 Distance를 새로운 오브젝트(도메인)으로 취급하고 Line, Station 연관 관계는 아예 다른 요구사항을 가진 관계로 풀어내면서 위에 느꼈던 문제들을 작은 문제로 나눠서 쉽게 해결되는 느낌을 받았습니다.

제가 느낀점이 리뷰어님이 의도하신 방향인지, 혹은 여전히 잘못된 방향이 있는지 가이드 주시면 감사하겠습니다 :)

young891221 reviewed (approved):
리뷰 반영해 주시느라 정말 고생많으셨습니다ㅎㅎ
제가 조금 더 잘 다듬어서 리뷰를 드렸어야 했는데 부족했네요ㅠㅠ
마지막 미션 잘 수행된것 같아 머지하겠습니다~!
